core/include/cubos/core/ecs/system.hpp

```diff
@@ -70,6 +70,162 @@ namespace cubos::core::ecs
     };
 
+    /// System argument which provides read access to the given resource or component.
+    /// Can be used as a pointer with both the `->` and `.` operators.
+    template <typename T>
+    class Read
+    {
+    public:
+        /// Creates a new read argument.
+        /// @param ref Reference to the resource or component.
+        inline Read(const T& ref)
+            : ref(ref)
+        {
+        }
+
+        /// @returns A pointer to the resource or component.
+        inline const T* operator->() const
+        {
+            return &this->ref;
+        }
+
+        /// @returns A reference to the resource or component.
+        inline const T& operator*() const
+        {
+            return this->ref;
+        }
+
+        /// @returns A reference to the resource or component.
+        inline const T& get() const
+        {
+            return this->ref;
+        }
+
+    private:
+        const T& ref; ///< Reference to the resource or component.
+    };
+
+    /// System argument which provides write access to the given resource or component.
+    /// Can be used as a pointer with both the `->` and `.` operators.
+    template <typename T>
+    class Write
+    {
+    public:
+        /// Creates a new write argument.
+        /// @param ref Reference to the resource or component.
+        inline Write(T& ref)
+            : ref(ref)
+        {
+        }
+
+        /// @returns A pointer to the resource or component.
+        inline T* operator->()
+        {
+            return &this->ref;
+        }
+
+        /// @returns A reference to the resource or component.
+        inline T& operator*()
+        {
+            return this->ref;
+        }
+
+        /// @returns A reference to the resource or component.
+        inline T& get()
+        {
+            return this->ref;
+        }
+
+    private:
+        T& ref; ///< Reference to the resource or component.
+    };
+
+    /// System argument which provides optional read access to the given resource or component, if it exists.
+    /// Can be used as a pointer with both the `->` and `.` operators.
+    template <typename T>
+    class OptRead
+    {
+    public:
+        /// Creates a new optional read argument.
+        /// @param ref Reference to the resource or component.
+        inline OptRead(const T* ref)
+            : ref(ref)
+        {
+        }
+
+        /// @returns A reference to the resource or component.
+        inline const T* operator->() const
+        {
+            return &this->get();
+        }
+
+        /// @returns A reference to the resource or component.
+        inline const T& operator*() const
+        {
+            return this->get();
+        }
+
+        /// @returns A reference to the resource or component.
+        inline T& get()
+        {
+            CUBOS_ASSERT(this->ref != nullptr, "Attempted to access a null optional resource or component");
+            return *this->ref;
+        }
```

Comment:
If these do the same why is their behaviour different?

Reply:
These are outdated because you moved the code in a later commit, the code is still de same in accessors.hpp

Reply:
> If these do the same why is their behaviour different?

Wdym?

Reply:
This was made irrelevant by removing the getter, disregard